Doc/c-api/monitoring.rst

```diff
@@ -0,0 +1,150 @@
+.. highlight:: c
+
+.. _monitoring:
+
+Monitorong C API
+================
+
+Added in version 3.13.
+
+An extension may need to interact with the event monitoring system. Subscribing
+to events and registering callbacks can be done via the Python API exposed in
+:mod:`sys.monitoring`.
+
+Generating Execution Events
+===========================
+
+The functions below make it possible for an extension to fire monitoring
+events as it emulates the execution of Python code. Each of these functions
+accepts a ``PyMonitoringState`` struct which contains concise information
+about the activation state of events, as well as the event arguments, which
+include a ``PyObject*`` representing the code object, the instruction offset
+and sometimes additional, event-specific arguments (see :mod:`sys.monitoring`
+for details about the signatures of the different event callbacks).
+The ``codelike`` argument should be an instance of :class:`types.CodeType`
+or of a type that emulates it.
+
+.. c:type:: PyMonitoringState
+
+  Representation of the state of an event type. It is allocated by the user
+  while its contents are maintained by the monitoring API functions described below.
+
+
+All of the functions below return 0 on success and -1 (with an exception set) on error.
```

Comment:
Mark said:
> Ideally, you shouldn't be calling monitoring functions with the exception set, except for those monitoring calls that are about exceptions. We clearly need to document this, though.

So I'd expect:


```suggestion
All of the functions below return 0 on success and -1 (with an exception set) on error.

They should not be called with an exception set, except for those that deal with
exceptions (those that fire ``RAISE``, ``C_RAISE``, ``RERAISE``, ``PY_UNWIND``, ``PY_THROW``, and ``STOP_ITERATION``).
```

Reply:
However, looking at the code/tests, the exception needs to be cleared:

```suggestion
All of the functions below return 0 on success and -1 (with an exception set) on error.

They should not be called with an exception set; you might want to wrap the
call in :c:func:`PyErr_GetRaisedException` and :c:func:`PyErr_SetRaisedException`.
```
... but it might be better to have that in the private `_PYMONITORING_IF_ACTIVE` blocks.